website/blog/2021-11-14-Presto-Cluster-Memory-Tuning.md

```diff
@@ -0,0 +1,70 @@
+---
+author: George Wang
+authorURL: https://www.linkedin.com/in/george-wang-9a5a46190/
+title: Presto Cluster Memory Tuning
+---
+
+##Presto Memory Management
+
+Presto's memory management system is carefully designed to achieve high performing query engine in a distributed system. It defines the memory into two categories: user memory and system memory.
+* The user memory is related to user data. For example, reading user data from cloud storage on AWS S3 bucket will occupy the corresponding memory. The amount of data is tightly related;
+* The system memory is meant for some memory allocated for the operator itself that needs to execute in order to process the query, such as tablescan operation and aggregation operation, for which the memory space may be required by operators, and are not strongly related to the user's data itself.
+
+Presto designs to divide the entire memory into three memory pools, namely System Pool, Reserved Pool, and General Pool. Reserved Pool and General Pool are used for query execution.  In general, the memory required for query execution is allocated from the general pool, and the reserved pool is usually idle meaning unused. When the general pool in the cluster is exhausted by a worker, the coordinator will select the query that occupies the most memory in the cluster, and allocate this query to the reserved pool, so that this large query can continue to execute by itself, and the freed memory will also make available for others query to continue. So that the entire system can move forward.
+
+###Why use memory pool
+
+The System Pool is used for the memory used by the system. For example, when transferring data between machines, the buffer is maintained in the memory. This part of the memory is mounted under the name of the system. So, why do you need reserved memory? If there is no Reserved Pool, when there are a lot of queries and the memory space is almost used up, a query that consumes a lot of memory starts to run. But at this time, there is no memory space for the query to run, and the query has been in a suspended state, waiting for available memory. But after other small-memory queries are finished, new small-memory queries are added. Because the small memory query occupies a small amount of memory, it is easy to find the available memory. In this case, the large-memory query will hang until starved to death, usually gets timed-out.
+
+Therefore, in order to prevent this kind of starvation, a space must be reserved for a large memory query to run. The reserved space is equal to the maximum memory allowed by the query. Every second, Presto picks out a query with the largest memory usage and allows it to use the reserved pool to avoid that there is always no available memory for the query to run.
```

Comment:
Presto only promotes queries to the reserved pool if there are blocked nodes on the cluster (nodes with full general pools)